src/main/java/uk/gov/hmcts/reform/et/syaapi/service/CaseDocumentService.java

```diff
@@ -0,0 +1,214 @@
+    private static final Pattern FILE_NAME_PATTERN = Pattern.compile(FILE_NAME_REGEX_PATTERN);
+    private static final String HTTPS_URL_REGEX_PATTERN =
+        "^(http://www\\.|https://www\\.|http://|https://)?[\\w\\-]+(\\.[a-z]{2,5})?(:\\d{1,5})?(/.*)?$";
+    private static final Pattern HTTPS_URL_PATTERN = Pattern.compile(HTTPS_URL_REGEX_PATTERN);
+    private static final String UPLOAD_FILE_EXCEPTION_MESSAGE = "Document management failed uploading file: ";
+    private static final String VALIDATE_FILE_EXCEPTION_MESSAGE = "File does not pass validation";
+    public final int maxApiRetries;
+    private final RestTemplate restTemplate;
+    private final AuthTokenGenerator authTokenGenerator;
+    private final String caseDocApiUrl;
+
+    /**
+     * Default constructor with injected parameters.
+     *
+     * @param restTemplate       the {@link RestTemplate} to be used to connect with the Case Document API
+     * @param authTokenGenerator the {@link AuthTokenGenerator} used to generate tokens for communicating with the
+     *                           Case Document API
+     * @param caseDocApiUrl      the URL to call the Case Document API
+     */
+    public CaseDocumentService(RestTemplate restTemplate,
+                               AuthTokenGenerator authTokenGenerator,
+                               @Value("${case_document_am.url}/cases/documents}")
+                                   String caseDocApiUrl,
+                               @Value("${CASE_DOCUMENT_AM_MAX_RETRIES}") Integer maxApiRetries) {
+        this.restTemplate = restTemplate;
+        this.authTokenGenerator = authTokenGenerator;
+        this.caseDocApiUrl = caseDocApiUrl;
+        this.maxApiRetries = maxApiRetries;
+    }
+
+    /**
+     * Given a file to upload, this will upload the file to the CCD document API and give back a unique ULR to access
+     * the uploaded file.
+     *
+     * @param authToken  the caller's bearer token used to verify the caller
+     * @param caseTypeId the area the file belongs to e.g. ET_EnglandWales
+     * @param file       the file to be uploaded
+     * @return the URL of the document we have just uploaded
+     * @throws CaseDocumentException if a problem occurs whilst uploading the document via API
+     */
+    public URI uploadDocument(String authToken, String caseTypeId, MultipartFile file) throws CaseDocumentException {
+        DocumentUploadResponse response = attemptWithRetriesToUploadDocumentToCaseDocumentApi(
+            0, authToken, caseTypeId, file);
+
+        CaseDocument caseDocument = validateResponse(
+            Objects.requireNonNull(response), file.getOriginalFilename());
+
+        return getUriFromFile(caseDocument, file.getOriginalFilename());
+    }
+
+    private URI getUriFromFile(CaseDocument caseDocument, String originalFilename) throws CaseDocumentException {
+        if (caseDocument.getLinks() == null
```

Comment:
Do these checks add any value? In what scenario will the links be null? Wouldn't it actually be better to allow a NPE to be thrown? The exception that is thrown seems to hide the problem

Reply:
This particular scenario check was implemented following feedback from @nickwest1, I specifically am not aware of any scenario.